### DISCUSSION TITLE
Can't do gracefull shutdown of apllication

### DISCUSSION THREAD

**Discussion by vokson** [emmett-framework/granian#295] [Q&A]
Granian 1.3.1

   My goal is to make program handling http request and send them to Kafka. I use `https://github.com/confluentinc/confluent-kafka-python` library, which is syncronious. Thus, I have to implement infinitive loop flushing kafka queue in separate thread for each granian's worker. The loop runs while flag `is_cancelled = False`.
   When I stop program (Ctrl-C, for example),
```python
def shutdown_app():
   is_cancelled = True

if __name__ == "__main__":
    try:
        app.serve()
    except KeyboardInterrupt:
        pass
    finally:
        shutdown_app()
```
I expect granian will stop and somehow move to shutdown_app() function to give me possibility to set `is_cancelled = True` and stop infinite thread. Unfortunately, granian never stops, because infinite thread is included in granian worker process.
```python
    def _stop_workers(self):
        for proc in self.procs:
            proc.terminate()

        for proc in self.procs:
            proc.join()  <<- HERE

        self.procs.clear()
```
If changed to `proc.join(1)', after 1 sec execution moves ahead and comes in `shutdown_app` function, then I can manage other stuff.

Possible solution:
 1. Add callback function to class Granian, which will called before stop of worker
 2. Add timeout property X to class Granian, which will be used as proc.join(X)

**Comment by gi0baro** [comment]
@vokson is that a WSGI application? For ASGI you can rely on the lifespan events to perform actions on shutdown.

Anyways, 1.4 will probably review signal handling and eventually add callbacks. In general, you can always customise what happens on the workers writing your own spawn method: for instance you can copy the code of `_spawn_wsgi_worker` (https://github.com/emmett-framework/granian/blob/master/granian/server.py#L284-L319) and add any code you want after the `serve` call (blocking til shutdown).

**Reply by vokson** [reply]
```
app = Granian(
    "main:main",
    interface=Interfaces.RSGI,
    loop=Loops.uvloop,
    address="0.0.0.0",
    port=PORT,
    workers=int(os.getenv("HTTP_PROCESSES", "1")),
    threads=int(os.getenv("HTTP_PROCESS_WORKERS", "1")),
)
```
I think, it is RSGI application. @gi0baro

**Reply by gi0baro** [reply]
@vokson understood.
RSGI might introduce also shutdown callbacks in the near future. In the meantime everything I wrote for `_spawn_wsgi_worker` is also true for `_spawn_rsgi_worker`.

**Reply by vokson** [reply]
Thank you. Let's wait shutdown callbacks functionality
